Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/publish.py

```diff
@@ -0,0 +1,255 @@
+        if args.remote not in repository.source_remotes():
+            sys.stderr.write("'{}' is not a recognized remote\n".format(args.remote))
+            return 1
+
+        existing_tags = set(repository.tags(remote=args.remote))
+        tags_to_publish = tags_to_publish - existing_tags
+        for commit in commits:
+            tags_to_publish |= set(cls.tags_on(repository, commit.hash)) - existing_tags
+            intersection = cls.parental_intersection(repository, commit)
+            if intersection and intersection.branch != repository.default_branch:
+                cls._add_branch_ref_to(
+                    mapping=branches_to_publish,
+                    repository=repository,
+                    branch=intersection.branch, commit=intersection,
+                )
+            for remote, branches in cls.branches_on(repository, commit.hash).items():
+                if remote is not None and remote not in repository.source_remotes():
+                    continue
+                for branch in branches:
+                    commit = repository.find('remotes/{}/{}'.format(remote, branch) if remote else branch)
+                    cls._add_branch_ref_to(
+                        mapping=branches_to_publish,
+                        repository=repository,
+                        branch=branch, commit=commit,
+                    )
+
+        if tags_to_publish - requested_refs:
+            log.warning("Inferred the following tags:")
+            for tag in tags_to_publish - requested_refs:
+                log.warning('    {}'.format(tag))
+            log.warning('')
+
+        if set(branches_to_publish.keys()) - requested_refs:
+            log.warning("Inferred the following branches:")
+            for branch in set(branches_to_publish.keys()) - requested_refs:
+                log.warning('    {} @ {}'.format(branch, branches_to_publish[branch]))
+            log.warning('')
+
+        if tags_to_publish:
+            print("Pushing the following tags to {}:".format(args.remote))
+            for tag in sorted(tags_to_publish):
+                print('    {}'.format(tag))
+            print('')
+        if branches_to_publish:
+            print("Pushing the branches at the following commits to {}:".format(args.remote))
+            for branch in sorted(branches_to_publish.keys()):
+                print('    {} @ {}'.format(branch, branches_to_publish[branch]))
+            print('')
+
+        if Terminal.choose(
+            "Are you sure you would like to publish to '{}' remote?".format(args.remote),
+            default='No',
+        ) == 'No':
+            sys.stderr.write("Publication canceled\n")
+            return 1
+
+        return_code = 0
+        for branch, commit in branches_to_publish.items():
+            print('Pushing {} to {}...'.format(branch, args.remote))
+            if run(
+                [repository.executable(), 'push', args.remote, '{}:refs/heads/{}'.format(commit.hash, branch)],
```

Comment:
Can we do a single atomic push here too?

Also, curious if you think there's value in showing the git commands we run here. Personally, I would appreciate seeing them for extra confidence that the tool is doing what I think it's doing.

Reply:
We can! Thought that we couldn't with branches, for some reason, but it does work.

I will add a logging statement showing the git commands being run, although that logging level isn't enabled by default.